src/borg/archive.py

```diff
@@ -820,34 +820,55 @@ def stat_attrs(self, st, path):
+        safe_path = make_path_safe(path)
+        item = Item(path=safe_path)
+        hardlink_master = False
+        hardlinked = st.st_nlink > 1
+        if hardlinked:
+            source = self.hard_links.get((st.st_ino, st.st_dev))
+            if source is not None:
+                item.source = source
+                status = 'h'  # hardlink (to already seen inodes)
+            else:
+                hardlink_master = True
+        yield item, status, hardlinked, hardlink_master
+        # if we get here, "with"-block worked ok without error/exception, the item was processed ok...
         self.add_item(item)
-        return 'd'  # directory
+        # ... and added to the archive, so we can remember it to refer to it later in the archive:
+        if hardlink_master:
+            self.hard_links[(st.st_ino, st.st_dev)] = safe_path
+
+    def process_dir(self, path, st):
+        with self.item_helper(path, st, 'd') as (item, status, hardlinked, hardlink_master):  # directory
+            item.update(self.stat_attrs(st, path))
+            return status
 
     def process_fifo(self, path, st):
-        item = Item(path=make_path_safe(path))
-        item.update(self.stat_attrs(st, path))
-        self.add_item(item)
-        return 'f'  # fifo
+        with self.item_helper(path, st, 'f') as (item, status, hardlinked, hardlink_master):  # fifo
+            item.update(self.stat_attrs(st, path))
+            return status
 
     def process_dev(self, path, st):
-        item = Item(path=make_path_safe(path), rdev=st.st_rdev)
-        item.update(self.stat_attrs(st, path))
-        self.add_item(item)
-        if stat.S_ISCHR(st.st_mode):
-            return 'c'  # char device
-        elif stat.S_ISBLK(st.st_mode):
-            return 'b'  # block device
+        with self.item_helper(path, st, None) as (item, status, hardlinked, hardlink_master):  # no status yet
+            item.rdev = st.st_rdev
+            item.update(self.stat_attrs(st, path))
+            if stat.S_ISCHR(st.st_mode):
+                return 'c'  # char device
+            elif stat.S_ISBLK(st.st_mode):
+                return 'b'  # block device
 
     def process_symlink(self, path, st):
-        with backup_io('readlink'):
-            source = os.readlink(path)
-        item = Item(path=make_path_safe(path), source=source)
-        item.update(self.stat_attrs(st, path))
-        self.add_item(item)
-        return 's'  # symlink
+        with self.item_helper(path, st, 's') as (item, status, hardlinked, hardlink_master):  # symlink
+            with backup_io('readlink'):
+                source = os.readlink(path)
+            item.source = source  # XXX this overwrites hardlink slave's usage of item.source
+            if hardlinked:
+                logger.warning('hardlinked symlinks will be extracted as non-hardlinked symlinks!')
+            item.update(self.stat_attrs(st, path))
+            return status
```

Comment:
In case someone wonders, yes, one can create hard links of device nodes, symlinks and fifos.

I'm just wondering whether it makes sense to really support that?